### PR TITLE
[worker, model] feat: add dense Qwen3 critic support to TorchTitan

### PR DESCRIPTION
**Pull request by zupengwang**
### What does this PR do?

Adds a dense Qwen3 critic to the TorchTitan engine so PPO/GAE can use the existing `value_model` worker and clipped value loss. Currently, selecting that combination fails with `Unknown backend: torchtitan`.

This implements the critic sub-item of [#5306](https://github.com/verl-project/verl/issues/5306), not the entire roadmap. A scalar head is created before model sharding and optimizer construction. HF backbone weights and optional `score.weight/score.bias` load through a dedicated adapter; native DCP retains the head, optimizer and scheduler.

### Checklist Before Starting

- [x] Checked issue #5306, its comments/timeline, and open/closed PRs. No matching critic implementation was found. Queries: [issue references](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+5306+in%3Abody), [TorchTitan critic](https://github.com/verl-project/verl/pulls?q=is%3Apr+torchtitan+critic).
- [x] Proposed title: `[worker, model] feat: add dense Qwen3 critic support to TorchTitan`.

### Test

Remote host: 2 × RTX 3090; Python 3.12.13; torch 2.14.0.dev20260625+cu130; torchtitan 0.1.0.dev20260701+cu130; transformers 5.9.0.

```bash
pytest -q tests/workers/test_torchtitan_value_model_on_cpu.py
torchrun --nproc_per_node=2 tests/special_distributed/test_torchtitan_critic.py --work-dir /tmp/critic-base
torchrun --nproc_per_node=2 tests/special_distributed/test_torchtitan_critic.py --work-dir /tmp/critic-score --with-score
pre-commit run --all-files --show-diff-on-failure
```

- 24 CPU tests passed.
- Both two-GPU cases passed, using a locally generated Qwen3 debug model and tokenizer.
- HF comparison maximum absolute error: 0.0087890625 from an LM checkpoint; 0.0068359375 from a checkpoint with a scalar score head.
- Both the head and backbone update with the existing clipped value loss.
- Model, optimizer and scheduler restore exactly. The next update after resume matches uninterrupted execution exactly.
- Existing actor forward regression matches the unmodified baseline.
- PPO/GAE Hydra config resolves to TorchTitanCriticConfig and TorchtitanEngineConfig.
- All 14 repository pre-commit hooks passed, and the commit hook passed again.

These tests cover the engine and existing PPO value loss. A complete rollout-backed PPO run, convergence and throughput are not claimed.

### API and Usage Example

`model_engine=torchtitan algorithm.adv_estimator=gae` uses the existing critic config. The documentation includes a complete set of overrides for the existing GSM8K script.

The initial implementation supports dense Qwen3 on CUDA with FSDP2 data parallelism. TP, CP, PP, EP and LoRA are explicitly rejected.

### Design & Code Changes

- Preserve vocabulary embeddings; use an untied scalar head initialized from HF `initializer_range`.
- Declare the bias sharding before TorchTitan parallelization.
- Omit only absent initial head parameters during HF loading; retain strict missing-backbone checks.
- Return differentiable token values in the original jagged layout for packed and padded execution.
- Select full FlexAttention for the critic. The supported dev20260625 nightly's short-query GQA decoding kernel failed the HF/packed-vs-padded parity check; switching only the attention kernel resolved the discrepancy.
- Reuse existing optimizer, value loss, and DCP lifecycle.
- Recognize the 2026 individual-contributor license header in the existing license checker.

### Checklist Before Submitting

- [x] Read AGENTS.md and CONTRIBUTING.md.
- [x] Human submitter reviewed the final change and authorized publication.
- [x] Relevant tests were executed on the remote GPU host; commands and results are listed above.
- [x] Add documentation and unit/distributed regression tests.
- [ ] Request upstream GPU CI after human review and publication. The existing TorchTitan workflow is currently under `.github/workflows/.stash/`; this patch does not activate it. CPU tests skip when TorchTitan is unavailable, and the distributed tests have been run manually on the remote host.
- [x] This PR contains only the reviewed commit `3f40ac0dda3404013f33edce0661c3f559c3c8f8`.

AI assistance: implementation, tests and this description were prepared with OpenAI Codex. The human submitter reviewed the final change and authorized publication. Upstream CI has not yet been requested.

### CODE REVIEW

File: verl/workers/engine/torchtitan/transformer_impl.py [hunk 1]
```diff
@@ -883,3 +894,79 @@ def forward_step(self, micro_batch: TensorDict, loss_function, forward_only):
...
+        # Do this before Trainer constructs/shards the model and its optimizer.
+        # Preserve the full input vocabulary and untie only the scalar head.
+        model_spec.model.enable_weight_tying = False
+        model_spec.model.lm_head.out_features = 1
+        model_spec.model.lm_head.bias = True
+        model_spec.model.lm_head.param_init = {
+            "weight": partial(torch.nn.init.normal_, std=self.model_config.hf_config.initializer_range),
+            "bias": torch.nn.init.zeros_,
+        }
+        model_spec.state_dict_adapter = Qwen3ValueStateDictAdapter
```
Comment (wuxibin89) [inline verl/workers/engine/torchtitan/transformer_impl.py:931]: TorchTitanEngineWithValueHead should be decoupled with specific model.

**Review by wuxibin89 — CHANGES_REQUESTED** [review]
